File: client/java/src/avt/MenuNPC.java
package avt;

import java.io.IOException;
import java.util.Vector;
import javax.microedition.lcdui.Graphics;
import javax.microedition.lcdui.Image;
import main.Canvas;

public final class MenuNPC extends MenuMain {
   private static MenuNPC me;
   private int idUser;
   private int x;
   private int y;
   private int w;
   private int h;
   private int xList;
   private int yList;
   private int wList;
   private int hList;
   private int hItem;
   private int focus;
   public static FrameImage imgDc;
   private Vector list = new Vector();
   private String nameNPC;
   private String[] textChat;
   private boolean[] isMenu;
   private int pa = 0;
   private int dyTran;
   private int timeOpen;
   private int pyLast;
   private boolean trans = false;
   private boolean isG = false;
   private long timeDelay;
   private long count;
   private long timePoint;
   private int vY;
   private int cmtoY;
   private int cmy;
   private int cmdy;
   private int cmvy;
   private int cmyLim;

   public static MenuNPC gI() {
      return me == null ? (me = new MenuNPC()) : me;
   }

   public MenuNPC() {
      this.w = 200 * AvMain.hd;
      this.h = 190 * AvMain.hd;
      this.x = (Canvas.w - this.w) / 2;
      this.y = (Canvas.h - this.h) / 2;
      this.yList = 70 * AvMain.hd;
      this.wList = 120 * AvMain.hd;
      this.xList = this.w - this.wList - 12 * AvMain.hd;
      this.hItem = 30 * AvMain.hd;
      this.hList = this.hItem * 3 + 20 * AvMain.hd;
      super.center = new Command(T.selectt, 0, this);
      super.right = new Command(T.close, 1, this);
   }

   public final void commandActionPointer(int var1) {
      switch (var1) {
         case 0:
            this.click();
            return;
         case 1:
            Canvas.menuMain = null;
         default:
      }
   }

   public final void setInfo(Vector var1, int var2, String var3, String var4, boolean[] var5) {
      this.list = var1;
      this.isMenu = var5;
      this.idUser = var2;
      this.cmyLim = var1.size() * this.hItem - (this.hList - 20 * AvMain.hd);
      if (this.cmyLim < 0) {
         this.cmyLim = 0;
      }

      this.nameNPC = var3;
      this.textChat = Canvas.M.splitFontBStrInLine(var4, this.w - 50 * AvMain.hd);
      Canvas.menuMain = this;
   }

   public final void update() {
      if (this.timeOpen > 0) {
         --this.timeOpen;
         if (this.timeOpen == 0) {
            this.click();
         }
      }

      if (this.vY != 0) {
         if (this.cmy < 0 || this.cmy > this.cmyLim) {
            this.vY -= this.vY / 4;
            this.cmy += this.vY / 20;
            if (this.vY / 10 <= 1) {
               this.vY = 0;
            }
         }

         if (this.cmy < 0) {
            if (this.cmy < -this.hList / 2) {
               this.cmy = -this.hList / 2;
               this.cmtoY = 0;
               this.vY = 0;
            }
         } else if (this.cmy > this.cmyLim) {
            if (this.cmy < this.cmyLim + this.hList / 2) {
               this.cmy = this.cmyLim + this.hList / 2;
               this.cmtoY = this.cmyLim;
               this.vY = 0;
            }
         } else {
            this.cmy += this.vY / 10;
         }

         this.cmtoY = this.cmy;
         this.vY -= this.vY / 10;
         if (this.vY / 10 == 0) {
            this.vY = 0;
         }
      } else if (this.cmy < 0) {
         this.cmtoY = 0;
      } else if (this.cmy > this.cmyLim) {
         this.cmtoY = this.cmyLim;
      }

      if (this.cmy != this.cmtoY) {
         this.cmvy = this.cmtoY - this.cmy << 2;
         this.cmdy += this.cmvy;
         this.cmy += this.cmdy >> 4;
         this.cmdy &= 15;
      }

   }

   public final void updateKey() {
      super.updateKey();
      ++this.count;
      boolean var1 = false;
      if (Canvas.a(2)) {
         --this.focus;
         if (this.focus < 0) {
            this.focus = this.list.size() - 1;
         }

         var1 = true;
      } else if (Canvas.a(8)) {
         ++this.focus;
         if (this.focus >= this.list.size()) {
            this.focus = 0;
         }

         var1 = true;
      }

      if (Canvas.isPointerClick) {
         this.pyLast = Canvas.pyLast;
         this.isG = false;
         if (Canvas.b(this.x + this.xList, this.y + this.yList, this.wList, this.hList)) {
            if (this.vY != 0) {
               this.isG = true;
            }

            Canvas.isPointerClick = false;
            this.pa = this.cmtoY;
            this.timeDelay = this.count;
            this.trans = true;
         }
      }

      if (this.trans) {
         int var2 = this.pyLast - Canvas.py;
         this.pyLast = Canvas.py;
         long var3 = this.count - this.timeDelay;
         int var5;
         int var6;
         if (Canvas.isPointerDown) {
            if (this.count % 2L == 0L) {
               this.dyTran = Canvas.py;
               this.timePoint = this.count;
            }

            this.vY = 0;
            if (Math.abs(var2) < 10 * AvMain.hd) {
               var5 = this.y + this.yList + 10 * AvMain.hd;
               var6 = this.hItem;
               if ((var5 = (this.cmtoY + Canvas.py - var5) / var6) >= 0 && var5 < this.list.size()) {
                  this.focus = var5;
               }
            }

            if (CRes.abs(Canvas.dy()) >= 10 * AvMain.hd) {
               super.isHide_ = true;
            } else if (var3 > 3L && var3 < 8L) {
               var5 = this.y + this.yList + 10 * AvMain.hd;
               var6 = this.hItem;
               if ((var5 = (this.cmtoY + Canvas.py - var5) / var6) >= 0 && var5 < this.list.size() && !this.isG) {
                  super.isHide_ = false;
               }
            }

            if (this.cmtoY < 0 || this.cmtoY > this.cmyLim) {
               this.cmtoY = this.pa + var2 / 2;
               this.pa = this.cmtoY;
            }

            this.cmy = this.cmtoY;
         }

         if (Canvas.isPointerRelease && Canvas.b(this.x, this.y, this.w, this.h)) {
            this.isG = false;
            var5 = (int)(this.count - this.timePoint);
            if (CRes.abs(var6 = this.dyTran - Canvas.py) > 40 && var5 < 10 && this.cmtoY > 0 && this.cmtoY < this.cmyLim) {
               this.vY = var6 / var5 * 10;
            }

            this.timePoint = -1L;
            if (Math.abs(var2) < 10 * AvMain.hd) {
               if (var3 <= 4L) {
                  super.isHide_ = false;
                  this.timeOpen = 5;
               } else if (!super.isHide_) {
                  this.click();
               }
            }

            this.trans = false;
            Canvas.isPointerRelease = false;
         }
      } else if (Canvas.isPointerRelease && !Canvas.b(this.x, this.y, this.w, this.h)) {
         Canvas.isPointerRelease = false;
         Canvas.menuMain = null;
      }

      if (var1) {
         this.cmtoY = this.focus * this.hItem - this.hList / 2 + this.hItem / 2;
         if (this.cmtoY > this.cmyLim) {
            this.cmtoY = this.cmyLim;
            return;
         }

         if (this.cmtoY < 0) {
            this.cmtoY = 0;
         }
      }

   }

   private void click() {
      if (!this.isMenu[this.focus]) {
         Canvas.menuMain = null;
      } else {
         Canvas.startWaitDlg();
      }

      ((Command)this.list.elementAt(this.focus)).perform();
   }

   public final void paint(Graphics var1) {
      Canvas.resetTrans(var1);
      Canvas.paint.paintPopupBack(var1, this.x, this.y, this.w, this.h, 0);
      var1.translate(this.x, this.y);
      var1.setColor(695195);
      var1.fillRect(12 * AvMain.hd, 12 * AvMain.hd, this.w - 24 * AvMain.hd, 50 * AvMain.hd);
      var1.setColor(12648440);
      var1.fillRect(15 * AvMain.hd, 15 * AvMain.hd, this.w - 30 * AvMain.hd, 44 * AvMain.hd);

      int var2;
      for(var2 = 0; var2 < this.textChat.length; ++var2) {
         Canvas.fontChatB.drawString(var1, this.textChat[var2], 20 * AvMain.hd, 12 * AvMain.hd + 25 * AvMain.hd - this.textChat.length * AvMain.hBlack / 2 + var2 * AvMain.hBlack, 0);
      }

      Avatar var9 = LoadMap.getAvatar(this.idUser);
      String npcName = this.nameNPC == null ? "" : this.nameNPC;
      Canvas.normalFont.drawString(var1, npcName, this.xList / 2, this.yList + this.hList / 2 - AvMain.hNormal - 20 * AvMain.hd, 2);
      if (var9 != null) {
         var9.paintIcon(var1, this.xList / 2, this.yList + this.hList / 2 + var9.height, true);
      }
      var2 = 4441283;
      FrameImage var7 = imgDc;
      int var6 = this.hList;
      int var5 = this.wList;
      int var4 = this.yList;
      int var3 = this.xList;
      Graphics var10 = var1;
      var7.drawFrame(0, var3, var4, 0, var1);
      var7.drawFrame(2, var3 + var5 - var7.frameWidth, var4, 0, var1);
      var7.drawFrame(5, var3, var4 + var6 - var7.frameHeight, 0, var1);
      var7.drawFrame(7, var3 + var5 - var7.frameWidth, var4 + var6 - var7.frameHeight, 0, var1);

      int var8;
      for(var8 = 0; var8 < (var5 - (var7.frameWidth << 1)) / var7.frameWidth; ++var8) {
         var7.drawFrame(1, var3 + (var8 + 1) * var7.frameWidth, var4, 0, var10);
         var7.drawFrame(6, var3 + (var8 + 1) * var7.frameWidth, var4 + var6 - var7.frameHeight, 0, var10);
      }

      var7.drawFrame(1, var3 + var5 - (var7.frameWidth << 1), var4, 0, var10);
      var7.drawFrame(6, var3 + var5 - (var7.frameWidth << 1), var4 + var6 - var7.frameHeight, 0, var10);

      for(var8 = 0; var8 < (var6 - (var7.frameHeight << 1)) / var7.frameHeight; ++var8) {
         var7.drawFrame(3, var3, var4 + (var8 + 1) * var7.frameHeight, 0, var10);
         var7.drawFrame(4, var3 + var5 - var7.frameWidth, var4 + (var8 + 1) * var7.frameHeight, 0, var10);
      }

      var7.drawFrame(3, var3, var4 + var6 - (var7.frameHeight << 1), 0, var10);
      var7.drawFrame(4, var3 + var5 - var7.frameWidth, var4 + var6 - (var7.frameHeight << 1), 0, var10);
      var10.setColor(4441283);
      var10.fillRect(var3 + var7.frameWidth, var4 + var7.frameHeight, var5 - (var7.frameWidth << 1), var6 - (var7.frameHeight << 1));
      var1.translate(this.xList, this.yList);
      var1.setClip(0, 0, this.wList, this.hList);
      var1.translate(0, -this.cmy);

      for(var2 = 0; var2 < this.list.size(); ++var2) {
         Command var11 = (Command)this.list.elementAt(var2);
         if (var2 == this.focus && !super.isHide_) {
            var1.setColor(10543802);
            var1.fillRect(4 * AvMain.hd, 10 * AvMain.hd + var2 * this.hItem, this.wList - 8 * AvMain.hd, this.hItem);
         }

         Canvas.normalFont.drawString(var1, var11.caption, 10 * AvMain.hd, 10 * AvMain.hd + var2 * this.hItem + this.hItem / 2 - AvMain.hNormal / 2, 0);
      }

      super.paint(var1);
   }

   static {
      try {
         imgDc = new FrameImage(Image.createImage(T.getPath() + "/race/popup/tile0.png"), 20 * AvMain.hd, 20 * AvMain.hd);
      } catch (IOException var1) {
         var1.printStackTrace();
      }

   }
}

File: client/unity/Assets/Scripts/Assembly-CSharp/MenuNPC.cs
public class MenuNPC : MenuMain
{
	public static MenuNPC me;

	private int idUser;

	private int x;

	private int y;

	private int w;

	private int h;

	private int xList;

	private int yList;

	private int wList;

	private int hList;

	private int hItem;

	private int focus;

	public static FrameImage imgDc;

	private MyVector list = new MyVector();

	private string nameNPC;

	private string[] textChat;

	private bool[] isMenu;

	private Image imgBackPopup;

	private Image imgBackChat;

	private int pa;

	private int dyTran;

	private int timeOpen;

	private int pyLast;

	private bool trans;

	private bool isG;

	private long timeDelay;

	private long count;

	private long timePoint;

	private int vY;

	public int cmtoY;

	public int cmy;

	public int cmdy;

	public int cmvy;

	public int cmyLim;

	static MenuNPC()
	{
		imgDc = new FrameImage(Image.createImagePNG(T.getPath() + "/race/popup/tile0"), 20 * AvMain.hd, 20 * AvMain.hd);
	}

	public MenuNPC()
	{
		w = 250 * AvMain.hd;
		h = 187 * AvMain.hd;
		x = (Canvas.w - w) / 2;
		y = (Canvas.h - h) / 2;
		yList = 75 * AvMain.hd;
		wList = 160 * AvMain.hd;
		xList = w - wList - 13 * AvMain.hd;
		hItem = 30 * AvMain.hd;
		hList = 100 * AvMain.hd;
	}

	public static MenuNPC gI()
	{
		return (me != null) ? me : (me = new MenuNPC());
	}

	public void setInfo(MyVector list, int idUser, string nameNPC, string textChat, bool[] isMenu)
	{
		if (imgBackPopup == null)
		{
			imgBackPopup = Image.createImagePNG(T.getPath() + "/effect/backMenuNPC");
			imgBackChat = Image.createImagePNG(T.getPath() + "/effect/popupChat");
		}
		this.list = list;
		this.isMenu = isMenu;
		this.idUser = idUser;
		cmyLim = list.size() * hItem - (hList - 20 * AvMain.hd);
		if (cmyLim < 0)
		{
			cmyLim = 0;
		}
		this.nameNPC = nameNPC;
		this.textChat = Canvas.fontChatB.splitFontBStrInLine(textChat, w - 70 * AvMain.hd);
		show();
	}

	public override void update()
	{
		if (timeOpen > 0)
		{
			timeOpen--;
			if (timeOpen == 0)
			{
				click();
			}
		}
		moveCamera();
	}

	public override void updateKey()
	{
		count++;
		if (Canvas.isPointerClick)
		{
			pyLast = Canvas.pyLast;
			isG = false;
			if (Canvas.isPoint(x + xList, y + yList, wList, hList))
			{
				if (vY != 0)
				{
					isG = true;
				}
				Canvas.isPointerClick = false;
				pa = cmtoY;
				timeDelay = count;
				trans = true;
			}
		}
		if (trans)
		{
			int num = pyLast - Canvas.py;
			pyLast = Canvas.py;
			long num2 = count - timeDelay;
			if (Canvas.isPointerDown)
			{
				if (count % 2 == 0)
				{
					dyTran = Canvas.py;
					timePoint = count;
				}
				vY = 0;
				if (Math.abs(num) < 10 * AvMain.hd)
				{
					int num3 = y + yList + 10 * AvMain.hd;
					int num4 = hItem;
					int num5 = (cmtoY + Canvas.py - num3) / num4;
					if (num5 >= 0 && num5 < list.size())
					{
						focus = num5;
					}
				}
				if (CRes.abs(Canvas.dy()) >= 10 * AvMain.hd)
				{
					isHide = true;
				}
				else if (num2 > 3 && num2 < 8)
				{
					int num6 = y + yList + 10 * AvMain.hd;
					int num7 = hItem;
					int num8 = (cmtoY + Canvas.py - num6) / num7;
					if (num8 >= 0 && num8 < list.size() && !isG)
					{
						isHide = false;
					}
				}
				if (cmtoY < 0 || cmtoY > cmyLim)
				{
					cmtoY = pa + num / 2;
					pa = cmtoY;
				}
				else
				{
					cmtoY = pa + num / 2;
					pa = cmtoY;
				}
				cmy = cmtoY;
			}
			if (!Canvas.isPointerRelease || !Canvas.isPoint(x, y, w, h))
			{
				return;
			}
			isG = false;
			int num9 = (int)(count - timePoint);
			int num10 = dyTran - Canvas.py;
			if (CRes.abs(num10) > 40 && num9 < 10 && cmtoY > 0 && cmtoY < cmyLim)
			{
				vY = num10 / num9 * 10;
			}
			timePoint = -1L;
			if (Math.abs(num) < 10 * AvMain.hd)
			{
				if (num2 <= 4)
				{
					isHide = false;
					timeOpen = 5;
				}
				else if (!isHide)
				{
					click();
				}
			}
			trans = false;
			Canvas.isPointerRelease = false;
		}
		else if (Canvas.isPointerRelease && !Canvas.isPoint(x, y, w, h))
		{
			Canvas.isPointerRelease = false;
			Canvas.menuMain = null;
		}
	}

	private void click()
	{
		if (!isMenu[focus])
		{
			Canvas.menuMain = null;
		}
		else
		{
			Canvas.startWaitDlg();
		}
		Command command = (Command)list.elementAt(focus);
		command.perform();
	}

	public override void paint(MyGraphics g)
	{
		Canvas.resetTrans(g);
		Canvas.paint.paintPopupBack(g, x, y, w, h, -1, false);
		g.translate(x, y);
		g.drawImage(imgBackChat, 12 * AvMain.hd, 12 * AvMain.hd, 0);
		for (int i = 0; i < textChat.Length; i++)
		{
			Canvas.fontChatB.drawString(g, textChat[i], 20 * AvMain.hd, 12 * AvMain.hd + 25 * AvMain.hd - textChat.Length * Canvas.fontChatB.getHeight() / 2 + i * AvMain.hBlack + ((AvMain.hd == 1) ? 10 : 0), 0);
		}
		Avatar avatar = LoadMap.getAvatar(idUser);
		Canvas.normalFont.drawString(g, nameNPC, xList / 2, yList + hList / 2 - AvMain.hNormal - 20 * AvMain.hd, 2);
		avatar.paintIcon(g, xList / 2, yList + hList / 2 + avatar.height, false);
		g.translate(xList, yList);
		g.drawImage(imgBackPopup, 0f, 0f, 0);
		g.setClip(0f, 0f, w, hList);
		g.translate(0f, -cmy);
		for (int j = 0; j < list.size(); j++)
		{
			Command command = (Command)list.elementAt(j);
			if (j == focus && !isHide)
			{
				g.setColor(16777215);
				g.fillRect(4 * AvMain.hd, 10 * AvMain.hd + j * hItem, wList - 6 * AvMain.hd, hItem);
			}
			Canvas.normalFont.drawString(g, command.caption, 10 * AvMain.hd, 10 * AvMain.hd + j * hItem + hItem / 2 - AvMain.hNormal / 2, 0);
		}
		base.paint(g);
	}

	public static void paintPopupTilte(MyGraphics g, int x, int y, int w, int h, FrameImage img, int color)
	{
		img.drawFrame(0, x, y, 0, g);
		img.drawFrame(2, x + w - img.frameWidth, y, 0, g);
		img.drawFrame(5, x, y + h - img.frameHeight, 0, g);
		img.drawFrame(7, x + w - img.frameWidth, y + h - img.frameHeight, 0, g);
		for (int i = 0; i < (w - img.frameWidth * 2) / img.frameWidth; i++)
		{
			img.drawFrame(1, x + (i + 1) * img.frameWidth, y, 0, g);
			img.drawFrame(6, x + (i + 1) * img.frameWidth, y + h - img.frameHeight, 0, g);
		}
		img.drawFrame(1, x + w - img.frameWidth * 2, y, 0, g);
		img.drawFrame(6, x + w - img.frameWidth * 2, y + h - img.frameHeight, 0, g);
		for (int j = 0; j < (h - img.frameHeight * 2) / img.frameHeight; j++)
		{
			img.drawFrame(3, x, y + (j + 1) * img.frameHeight, 0, g);
			img.drawFrame(4, x + w - img.frameWidth, y + (j + 1) * img.frameHeight, 0, g);
		}
		img.drawFrame(3, x, y + h - img.frameHeight * 2, 0, g);
		img.drawFrame(4, x + w - img.frameWidth, y + h - img.frameHeight * 2, 0, g);
		if (color != -1)
		{
			g.setColor(color);
			g.fillRect(x + img.frameWidth, y + img.frameHeight, w - img.frameWidth * 2, h - img.frameHeight * 2);
		}
	}

	public void moveCamera()
	{
		if (vY != 0)
		{
			if (cmy < 0 || cmy > cmyLim)
			{
				vY -= vY / 4;
				cmy += vY / 20;
				if (vY / 10 <= 1)
				{
					vY = 0;
				}
			}
			if (cmy < 0)
			{
				if (cmy < -hList / 2)
				{
					cmy = -hList / 2;
					cmtoY = 0;
					vY = 0;
				}
			}
			else if (cmy > cmyLim)
			{
				if (cmy < cmyLim + hList / 2)
				{
					cmy = cmyLim + hList / 2;
					cmtoY = cmyLim;
					vY = 0;
				}
			}
			else
			{
				cmy += vY / 10;
			}
			cmtoY = cmy;
			vY -= vY / 10;
			if (vY / 10 == 0)
			{
				vY = 0;
			}
		}
		else if (cmy < 0)
		{
			cmtoY = 0;
		}
		else if (cmy > cmyLim)
		{
			cmtoY = cmyLim;
		}
		if (cmy != cmtoY)
		{
			cmvy = cmtoY - cmy << 2;
			cmdy += cmvy;
			cmy += cmdy >> 4;
			cmdy &= 15;
		}
	}
}
